File: middleware/communicator/lapras_topic.py
from typing import List, Optional
from communicator.mqtt_topic import MqttTopic
from communicator.message_type import MessageType

class LaprasTopic(MqttTopic):
    def __init__(self, place_name: Optional[str], message_type: Optional[MessageType], *subtopics):
        super().__init__()
        self.tokens = [None] * (2 + len(subtopics))
        self.set_place_name(place_name)
        self.set_message_type(message_type)
        self.set_subtopics(*subtopics)
    
    @classmethod
    def from_string(cls, topic_string: str) -> 'LaprasTopic':
        topic = cls(None, None)
        topic.tokens = topic_string.split("/")
        return topic
    
    def get_place_name(self) -> Optional[str]:
        return self.tokens[0]
    
    def set_place_name(self, place_name: Optional[str]):
        self.tokens[0] = place_name
    
    def get_message_type(self) -> Optional[MessageType]:
        if self.tokens[1] is None:
            return None
        return MessageType.from_string(self.tokens[1])
    
    def set_message_type(self, message_type: Optional[MessageType]):
        if message_type is None:
            self.tokens[1] = None
        else:
            self.tokens[1] = str(message_type)
    
    def get_subtopics(self) -> List[str]:
        return self.tokens[2:]
    
    def set_subtopics(self, *subtopics):
        for i, subtopic in enumerate(subtopics):
            if i < len(self.tokens) - 2:
                self.tokens[i + 2] = subtopic

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/communicator/LaprasTopic.java
package kr.ac.kaist.cdsn.lapras.communicator;

import java.util.Arrays;

/**
 * Created by Daekeun Lee on 2016-11-15.
 */
public class LaprasTopic extends MqttTopic {
    public LaprasTopic(String placeName, MessageType messageType, String... subtopics) {
        tokens = new String[2 + subtopics.length];
        setPlaceName(placeName);
        setMessageType(messageType);
        setSubtopics(subtopics);
    }

    public LaprasTopic(String topicString) {
        super(topicString);
    }

    public String getPlaceName() {
        return tokens[0];
    }

    public void setPlaceName(String placeName) {
        tokens[0] = placeName;
    }

    public MessageType getMessageType() {
        return MessageType.fromString(tokens[1]);
    }

    public void setMessageType(MessageType messageType) {
        tokens[1] = (messageType == null) ? null : messageType.toString();
    }

    public String[] getSubtopics() {
        return Arrays.copyOfRange(tokens, 2, tokens.length);
    }

    public void setSubtopics(String... subtopics) {
        System.arraycopy(subtopics, 0, tokens, 2, subtopics.length);
    }
}
